src/os/bluestore/BlueStore.h

```diff
@@ -490,39 +497,190 @@ class BlueStore : public ObjectStore,
 
   /// a cache (shard) of onodes and buffers
   struct Cache {
+    std::mutex lock;                ///< protect lru and other structures
+
+    static Cache *create(string type);
+
+    virtual ~Cache() {}
+
+    virtual void _add_onode(OnodeRef& o, int level) = 0;
+    virtual void _rm_onode(OnodeRef& o) = 0;
+    virtual void _touch_onode(OnodeRef& o) = 0;
+
+    virtual void _add_buffer(Buffer *b, int level, Buffer *near) = 0;
+    /// remove a buffer.  return level to be used by any replacement buffer.
+    virtual int _rm_buffer(Buffer *b) = 0;
+    virtual void _adjust_buffer_size(Buffer *b, int64_t delta) = 0;
+    virtual void _touch_buffer(Buffer *b) = 0;
+
+    virtual void trim(uint64_t onode_max, uint64_t buffer_max) = 0;
+
+#ifdef DEBUG_CACHE
```

Comment:
IMHO it's better to move DEBUG_CACHE definition to Bluestore.cc and fork implementation there only. And let interfaces to be the same for both defined and undefined cases. No much overhead in runtime if any. And the code is more readable and no huge rebuilds if there is a need to change.